dm/syncer/dml.go

```diff
@@ -683,14 +685,34 @@ func genKeyList(table string, columns []*model.ColumnInfo, dataSeq []interface{}
 	return buf.String()
 }
 
+// truncateIndexValues truncate prefix index from data.
+func truncateIndexValues(indexColumns *model.IndexInfo, tiColumns []*model.ColumnInfo, data []interface{}) []interface{} {
+	values := make([]interface{}, 0, len(indexColumns.Columns))
+	for i, iColumn := range indexColumns.Columns {
+		tcolumn := tiColumns[i]
+		if data[i] != nil && (tcolumn.Tp == mysql.TypeVarchar || tcolumn.Tp == mysql.TypeVarString ||
+			tcolumn.Tp == mysql.TypeString || tcolumn.Tp == mysql.TypeBlob || tcolumn.Tp == mysql.TypeTinyBlob ||
+			tcolumn.Tp == mysql.TypeMediumBlob || tcolumn.Tp == mysql.TypeLongBlob) {
+			datum := ttypes.NewDatum(data[i])
+			tablecodec.TruncateIndexValue(&datum, iColumn, tcolumn)
+			values = append(values, datum.GetValue())
+		} else {
+			values = append(values, data[i])
+		}
+	}
+	return values
+}
+
 // genMultipleKeys gens keys with UNIQUE NOT NULL value.
 // if not UNIQUE NOT NULL value, use table name instead.
 func genMultipleKeys(downstreamTableInfo *schema.DownstreamTableInfo, ti *model.TableInfo, value []interface{}, table string) []string {
 	multipleKeys := make([]string, 0, len(downstreamTableInfo.AvailableUKIndexList))
 
 	for _, indexCols := range downstreamTableInfo.AvailableUKIndexList {
 		cols, vals := getColumnData(ti.Columns, indexCols, value)
-		key := genKeyList(table, cols, vals)
+		// handle prefix index
+		truncVals := truncateIndexValues(indexCols, cols, vals)
```

Comment:
do we need to do some extra logic fot other index type? such as  func_index https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-functional-key-parts

Reply:
I will test func_index later, but I'm afraid `TruncateIndexValue` does not support.

Reply:
it seems we do not support function index.
scheam: `CREATE TABLE t1 (col1 INT, col2 INT, unique key func_index ((abs(col1))));`
the function index will be a column `_V$_func_index_0` with offset `2`, and we will have a wrong uk which column offset is wrong.

Reply:
we can solve this in another pr, so we need keep https://github.com/pingcap/ticdc/issues/3286 open

Reply:
yes, it seems that function index is different from prefix index, and we will support it in another pr.